### ISSUE TITLE
How to use this repo to download m3u8 live stream URL and save to local hard drive?

### ISSUE THREAD

**Issue by zydjohnHotmail** [SuRGeoNix/Flyleaf#47]
Hello:
I happen to find your repo, and it seems to be interesting. However, as it is rather complicated, 
I can’t figure out yet how to download multiple M3U8 videos and save to local hard drive.
For example, I have the following 2 M3U8 videos in the URL:
http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/.m3u8
http://devimages.apple.com/iphone/samples/bipbop/bipbopall.m3u8
I want to launch 2 clients to download them at the same time, when finished, I want to save the files as local files: C:\1.mp4 and C:\2.mp4 and stop or dispose the client.
Please show me how I can do this with C#.
By the way, I am using Windows 10 Version 21H1, and Visual Studio 2019 Version 16.10.4.
Thanks,

**Comment by SuRGeoNix** [comment]
Hopefully, I will find sometime to write a documentation soon. Until then, I have some samples to help you out with the basic implementations. For the downloader, I have this one https://github.com/SuRGeoNix/Flyleaf/blob/master/Wpf%20Samples/Sample4_Downloader.xaml.cs

Did you have a look and you still need some help?

**Comment by zydjohnHotmail** [comment]
Hello:
Thank you very much for your reply. 
I have some concerns: first, I don't know how to call this partical class, do you have the code sample for the caller?
second, I don't like to use WPF, which has many restrictions, can I use the class in either C# console project or C# WindForm project?
finally, I see your code:
public DemuxerBase Downloader { get; set; } = new VideoDemuxer(new Config(), 777);
I want to know what the number 777 means.
In my case, I have to download multiple M3U8 live streaming URLs at the same time.  For example, I have 2 Olympic games are on going at the same time.
Can I use one client with Config() ID# 777, and another client with Config() ID# 888?
In short, I want to initalize the class instance from the caller with different Config() ID#, and different M3U8 URL.
Can this be done in C# console or WinForm project?
Finally, the download percentage can NOT decided for real live streaming URL.
I give you one of such URL, it will expire in a few hours from now.
https://video1.maxline.by:444/o1/stream04/chunks.m3u8?nimblesessionid=3025892
Can you show me the code on how to reach my target?
Thanks,

**Comment by SuRGeoNix** [comment]
Hey @zydjohnHotmail 

To use the solution right click on Wpf Samples project and select "Set as Startup Project". Then edit App.xaml and set the StartupUri to Sample4_Downloader.xaml. That's all.

Now, the code that you ask is already inside that project, the library code can be used from any project that you reference (winforms/wpf/consoles).

Basic steps to use the library for this purpose:

1) Register FFmpeg libraries

```
Master.RegisterFFmpeg(@"c:\ffmpeg\x64");
```

2) Create the Demuxer/Downloader classes (yes for each you need a unique id until I will decide what to do with it)

```
DemuxerBase Downloader = new VideoDemuxer(new Config(), 1);
```

3) Register to event

```
Downloader.DownloadCompleted += Downloader_DownloadCompleted;
```

4) Open input url

```
if (Downloader.Open(url) != 0) { MessageBox.Show("Could not open url input");  return; }
```

5) Select the audio/video streams that you want to download (validate that they exist and what quality you want)

```
Downloader.EnableStream(Downloader.VideoStreams[0]);
if (Downloader.AudioStreams.Count != 0) Downloader.EnableStream(Downloader.AudioStreams[0]);
```

6) Start the downloader

```
Downloader.Download(System.IO.Path.Combine(System.IO.Path.GetTempPath(), $"testVideo{downloadCounter++}.mp4"));
```

7) Catch the complete event

```
private void Downloader_DownloadCompleted(object sender, bool success)
{
    if (success)
    {
        if (Downloader.DownloadPercentage == 100)
            MessageBox.Show("Download Completed!");
        else
            MessageBox.Show("Partial/Live Download Completed!");
    }
    else
        MessageBox.Show("Download Failed!");
}
```

**Comment by zydjohnHotmail** [comment]
Hello:
Thank you very much for your kind help.
I tried to follow your advice, and I can download fixed size video. But for the live streaming video, I see that the program download only first 30 seconds, then it stops. I don't know how to fix this.
You can try this URL: 
https://video1.maxline.by:444/o1/stream30/chunks.m3u8?nimblesessionid=3027958
But it will expire in a few hours.
Thanks,

**Comment by SuRGeoNix** [comment]
You are right, I'm looking at it

**Comment by SuRGeoNix** [comment]
It should be fixed now. Have a look and let me know if not. You might also want to check the timeouts/options/flags etc. for more advanced things.. see [Demuxer's Config](https://github.com/SuRGeoNix/Flyleaf/blob/master/FlyleafLib/Config.cs), [FFmpeg's protocols](https://ffmpeg.org/ffmpeg-protocols.html), [FFmpeg's formats](https://ffmpeg.org/ffmpeg-formats.html)

eg. you might need to use this for live streams:

```
https://ffmpeg.org/ffmpeg-formats.html
```
